pkg/reconciler/instances/istio/reset/pod/handler.go

```diff
@@ -112,10 +134,162 @@ func (i *RolloutHandler) ExecuteAndWaitFor(context context.Context, object Custo
 	}, i.retryOpts...)
 
 	if err != nil {
+		i.log.Warnf("unable to rollout %s/%s/%s: %v", object.Kind, object.Namespace, object.Name, err)
+		errAnnotate := k8sRetry.RetryOnConflict(k8sRetry.DefaultRetry, func() error {
+			client := i.kubeClient
+			err := annotateObject(context, client, object, AnnotationResetWarningKey, AnnotationResetWarningRolloutTimeoutVal)
+			return err
+		})
+
+		if errAnnotate != nil {
+			return errAnnotate
+		}
+		return err
+	}
+
+	err = i.WaitForResources(context, object)
+	if err != nil {
+		errAnnotate := k8sRetry.RetryOnConflict(k8sRetry.DefaultRetry, func() error {
+			client := i.kubeClient
+			err := annotateObject(context, client, object, AnnotationResetWarningKey, AnnotationResetWarningRolloutTimeoutVal)
+			return err
+		})
+
+		if errAnnotate != nil {
+			return errAnnotate
+		}
 		return err
 	}
 
-	return i.WaitForResources(context, object)
+	return nil
+}
+
+func annotateObject(context context.Context, client kubernetes.Interface, object CustomObject, key string, val string) error {
+	switch object.Kind {
+	case "Deployment":
+		err := annotatePodsFor(context, getDeploymentLabelSelector, client, object, key, val)
```

Comment:
I can see a repetitive pattern in this function. For every k8s object you call `annotatePodsFor(...)`. What about creating a dedicated function around switch-case that would simply return us a label selector and then call mentioned function once for returned argument? Wouldn't it be cleaner?